src/daemon/distribution.c

```diff
@@ -0,0 +1,280 @@
+  return node_index + 1;
+}
+
+static size_t right_child_index(size_t node_index, size_t left, size_t right) {
+  size_t mid = (left + right) / 2;
+  return node_index + 2 * (mid - left + 1);
+}
+
+static size_t tree_size(size_t num_buckets) {
+  return 2 * num_buckets - 1;
+}
+
+static bucket_t merge_buckets(bucket_t left_child, bucket_t right_child) {
+  return (bucket_t) {
+      .bucket_counter = left_child.bucket_counter + right_child.bucket_counter,
+      .maximum = right_child.maximum,
+  };
+}
+
+static void build_tree(distribution_t *d, bucket_t *buckets, size_t node_index, size_t left, size_t right) {
+  if (left > right)
+    return;
+  if (left == right) {
+    d->tree[node_index] = buckets[left];
+    return;
+  }
+  size_t mid = (left + right) / 2;
+  size_t left_child = left_child_index(node_index, left, right);
+  size_t right_child = right_child_index(node_index, left, right);
+  build_tree(d, buckets, left_child, left, mid);
+  build_tree(d, buckets, right_child, mid + 1, right);
+  d->tree[node_index] = merge_buckets(d->tree[left_child], d->tree[right_child]);
+}
+
+static distribution_t* build_distribution_from_bucket_array(size_t num_buckets, bucket_t *bucket_array) {
+  distribution_t *new_distribution = calloc(1, sizeof(*new_distribution));
+  if (num_buckets == 0)
+    return new_distribution;
+  bucket_t *nodes = calloc(tree_size(num_buckets), sizeof(*nodes));
+  if (new_distribution == NULL || nodes == NULL) {
+    free(new_distribution);
+    free(nodes);
+    return NULL;
+  }
+  new_distribution->tree = nodes;
+
+  new_distribution->num_buckets = num_buckets;
+  build_tree(new_distribution, bucket_array, 0, 0, num_buckets - 1);
+  return new_distribution;
+}
+
+distribution_t* distribution_new_linear(size_t num_buckets, double size) {
+  if (num_buckets == 0 || size <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  bucket_t bucket_array[num_buckets];
+  for (size_t i = 0; i < num_buckets; i++) {
+    bucket_array[i] = (bucket_t) {
+        .bucket_counter = 0,
+        .maximum = (i == num_buckets - 1) ? INFINITY : (i + 1) * size,
+    };
+  }
+  return build_distribution_from_bucket_array(num_buckets, bucket_array);
+}
+
+distribution_t* distribution_new_exponential(size_t num_buckets, double base, double factor) {
+  if (num_buckets == 0 || base <= 1 || factor <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  bucket_t bucket_array[num_buckets];
+  for (size_t i = 0; i < num_buckets; i++) {
+    bucket_array[i] = (bucket_t) {
+        .bucket_counter = 0,
+        .maximum = (i == num_buckets - 1) ? INFINITY : factor * pow(base, i), //check if it's slow
+    };
+  }
+  return build_distribution_from_bucket_array(num_buckets, bucket_array);
+}
+
+distribution_t* distribution_new_custom(size_t array_size, double *custom_buckets_boundaries) {
+  for (size_t i = 0; i < array_size; i++) {
+    double previous_boundary = 0;
+    if (i > 0)
+      previous_boundary = custom_buckets_boundaries[i - 1];
```

Comment:
For consideration: 
surrounding even the short statement with brackets would make the code more readable

Reply:
I would also consider adding brackets here for better readability.

Reply:
Thanks for your suggestions! I'll do this